src/set_ref.rs

```diff
@@ -0,0 +1,229 @@
+use crate::iter::*;
+use crate::{HashSet, GuardRef, TryInsertError};
+use crossbeam_epoch::Guard;
+use std::borrow::Borrow;
+use std::fmt::{self, Debug, Formatter};
+use std::hash::{BuildHasher, Hash};
+use std::ops::Index;
+
+/// A reference to a [`HashSet`], constructed with [`HashSet::pin`] or [`HashSet::with_guard`].
+///
+/// The current thread will be pinned for the duration of this reference.
+/// Keep in mind that this prevents the collection of garbage generated by the set.
+pub struct HashSetRef<'set, T, S = crate::DefaultHashBuilder> {
+    set: &'set HashSet<T, S>,
+    guard: GuardRef<'set>,
+}
+
+impl<K, V, S> HashSet<T, S> {
+    /// Get a reference to this set with the current thread pinned.
+    ///
+    /// Keep in mind that for as long as you hold onto this, you are preventing the collection of
+    /// garbage generated by the set.
+    pub fn pin(&self) -> HashSetRef<'_, T, S> {
+        HashSetRef {
+            guard: GuardRef::Owned(self.guard()),
+            set: &self,
+        }
+    }
+
+    /// Get a reference to this set with the given guard.
+    pub fn with_guard<'g>(&'g self, guard: &'g Guard) -> HashSetRef<'g, T, S> {
+        HashSetRef {
+            set: &self,
+            guard: GuardRef::Ref(guard),
+        }
+    }
+}
+
+impl<T, S> HashSetRef<'_, T, S> {
+    /// An iterator visiting all key-value pairs in arbitrary order.
+    /// The iterator element type is `(&'g K, &'g V)`.
+    /// See also [`HashSet::iter`].
+    pub fn iter(&self) -> Keys<'_, T, ()> {
```

Comment:
We should do the same re-org for this as I recently did for `map_ref` so that the methods and impl blocks are all in the same order as they are in `set.rs`.

Reply:
Alright. were should I place it then?

Reply:
Actually, how about I just do the re-ordering? Might be faster, since I know what I want :p I _think_ I can just push to your branch.